### ISSUE TITLE
Heavily refactor CircleCI config

### ISSUE THREAD

**Issue by AntonyMilneQB** [kedro-org/kedro-viz#812]
Our CircleCI config is looking pretty ropey. I think we should do a refactor similar to the changes I made to the kedro framework CircleCI config a few months ago: https://github.com/kedro-org/kedro/pull/1102 https://github.com/kedro-org/kedro/pull/1107

Some particular pain points with the current setup - do feel free to add if you know of any more:
- it's not easy to add new Python versions to the build (would be fixed by using `matrix` as in the PRs linked to above). Hence it's hard to rigorously test compatibility with Python 3.9 and 3.10 or assess the impact of bumping dependencies
- we [temporarily disabled all builds except Python 3.8 nearly 18 months ago](https://github.com/kedro-org/kedro-viz/pull/305) and never put them back... 😅 
- no separation of linting and testing into separate workflows (a minor annoyance but easy to fix)
- unclear exactly how it works, e.g. is `CONDA_ENV_PY_VERSION` doing anything?
- hard to follow due to the number of layers of indirection

Overall I think we could make a substantial improvement to our setup without that much effort, and I don't think the additional workflows will slow it down much either since they should run in parallel.

**Comment by AntonyMilneQB** [comment]
Slightly different from just refactoring, but at the same time I think it's worth doing the following also:
- convert scheduled workflows into scheduled pipelines since scheduled workflows will be deprecated later this year. See https://github.com/kedro-org/kedro/issues/1319
- consider whether our environment setup (e.g. caching, whether we use a custom docker image, how we install dependencies) could be improved - as done for Windows in https://github.com/kedro-org/kedro/pull/1342
- make sure we're running the test in parallel as much as possible to give quickest CI run times

**Comment by tynandebold** [comment]
Part 1 is complete. Part 2 is here: #912
